File: python3/partition_list.py
from .model import ListNode

"""
Space   : O(n)
Time    : O(n)
"""


class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        if not head:
            return head

        p = head
        a = ListNode(0)  # >= x
        p1 = a
        b = ListNode(0)  # < x
        p2 = b

        while p:
            if p.val >= x:
                p1.next = ListNode(p.val)
                p1 = p1.next
            else:
                p2.next = ListNode(p.val)
                p2 = p2.next
            p = p.next

        p2 = b
        while p2:
            if not p2.next:
                p2.next = a.next
                break
            p2 = p2.next

        return b.next

File: golang/partition_list.go
package main

/*
Space	: O(1)
Time	: O(n)
*/

import "golang/model"

func partition(head *model.ListNode, x int) *model.ListNode {
	gte, lt := &model.ListNode{}, &model.ListNode{}
	p1, p2 := gte, lt

	cur := head
	for cur != nil {
		if cur.Val >= x {
			p1.Next = cur
			p1 = cur
			cur = cur.Next
			p1.Next = nil
		} else {
			p2.Next = cur
			p2 = cur
			cur = cur.Next
			p2.Next = nil
		}
	}

	p2.Next = gte.Next
	return lt.Next
}
